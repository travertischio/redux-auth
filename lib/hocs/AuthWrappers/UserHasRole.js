'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _includes2 = require('lodash/includes');

var _includes3 = _interopRequireDefault(_includes2);

var _reduxAuthWrapper = require('redux-auth-wrapper');

var _reactRouterRedux = require('react-router-redux');

var _selectors = require('redux-auth/lib/containers/AuthenticationProvider/selectors');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var UserHasRole = function UserHasRole(expectedRoles) {
  return (0, _reduxAuthWrapper.UserAuthWrapper)({
    authSelector: _selectors.selectUser,
    predicate: getUserHasOfEpectedRoleFn(expectedRoles),
    redirectAction: _reactRouterRedux.routerActions.replace,
    // TODO: move url of default homepage/dashboard to the config file
    failureRedirectPath: '/home',
    allowRedirectBack: false,
    wrapperDisplayName: 'UserHasRole'
  });
};

function getUserHasOfEpectedRoleFn(expectedRoles) {
  return function (user) {
    return user && (0, _includes3.default)(expectedRoles, user.role);
  };
}

exports.default = UserHasRole;