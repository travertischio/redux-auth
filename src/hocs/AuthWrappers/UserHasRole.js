import { redirectActionWithSupportParamInQueryString } from '~/containers/AuthenticationProvider/actions';
import { selectUser } from '~/containers/AuthenticationProvider/selectors';
import config from '~/config';
import connectedReduxRedirect from './connectedReduxRedirect';

const UserHasRole = (expectedRoles) => connectedReduxRedirect({
  authenticatedSelector: makeUserHasOfEpectedRoleSelector(expectedRoles),
  redirectAction: redirectActionWithSupportParamInQueryString,
  redirectPath: () => config.userHasNoRoleRedirectPath,
  allowRedirectBack: false,
  wrapperDisplayName: 'UserHasRole',
});

function makeUserHasOfEpectedRoleSelector(expectedRoles) {
  return (state) => {
    const user = selectUser(state);

    return user && expectedRoles.includes(user.role);
  };
}

export default UserHasRole;
