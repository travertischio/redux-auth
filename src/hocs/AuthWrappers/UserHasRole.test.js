import React from 'react';
import { createComponentWithRouter } from 'react-unit-testing-utils';
import UserHasRole from './UserHasRole';

const testRoles = ['20_editor', '30_manager'];
const PageComponent = () => <div>Page only for users with specific roles</div>;
const PageOnlyForUsersWithRoles = UserHasRole(testRoles)(PageComponent);

describe('UserHasRole', () => {
  it('should be defined', () => {
    expect(PageOnlyForUsersWithRoles).toBeDefined();
  });

  it('should redner PageComponent when user has "20_editor" role', () => {
    const initialState = {
      auth: {
        user: {
          role: '20_editor',
        },
      },
    };

    const { wrapper } = createComponentWithRouter(PageOnlyForUsersWithRoles, initialState);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should redner PageComponent when user has "30_manager" role', () => {
    const initialState = {
      auth: {
        user: {
          role: '30_manager',
        },
      },
    };

    const { wrapper } = createComponentWithRouter(PageOnlyForUsersWithRoles, initialState);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should not render PageComponent when user has different role that specified', () => {
    const initialState = {
      auth: {
        user: {
          role: '10_user',
        },
      },
    };

    const { wrapper } = createComponentWithRouter(PageOnlyForUsersWithRoles, initialState);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should not render PageComponent when user is not authenticated', () => {
    const initialState = {
      auth: {},
    };

    const { wrapper } = createComponentWithRouter(PageOnlyForUsersWithRoles, initialState);

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
