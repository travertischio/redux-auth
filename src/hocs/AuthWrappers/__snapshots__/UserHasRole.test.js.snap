// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserHasRole should not render PageComponent when user has different role that specified 1`] = `<main />`;

exports[`UserHasRole should not render PageComponent when user is not authenticated 1`] = `<main />`;

exports[`UserHasRole should redner PageComponent when user has "20_editor" role 1`] = `
<main>
  <div>
    Page only for users with specific roles
  </div>
</main>
`;

exports[`UserHasRole should redner PageComponent when user has "30_manager" role 1`] = `
<main>
  <div>
    Page only for users with specific roles
  </div>
</main>
`;
